gq/gq_iso_test.go

```diff
@@ -0,0 +1,162 @@
+const vHex = "010000000000000000000D"
+const qHex = "3BED38CEBB1219BC068774E0E2655CDEF67FE547BCF2D9FA9FE167B1E63B2F101A1483D38A8F24EDE365A3E44F4F10ADECEA7B30D042C14C162477B8184AE6CFAA78441B1FDFB0B223ABCD528B61F313D859FCF9C26FCAF9E4D9DA9BA83E9D2FDA041E8CCBF90056C31D654B546C1A7F6729A8DD8E68512F39E3B6F07959CE61"
+const idHex = "416C657820416D706C65"
+const gHex = "3E641A22D0D0747D4ACC71884D3DFF2B2ADFDC1703B5A74EFD8333AB8C4377BB2A9B48E707F73409ABFBCD2DED69F52B16A145CE062FE6BD712C1952110DFB2316C5F3F321922ED375A4DEB8C41FA79BCAD86B0EA0D8FF02C9D0D5911BFF1E87DBCF073F71F18C08EB944AE84883A1E13FB1DEA123B5B1EFEA2A92635BD5D88F"
+const rHex = "487CDB0041BEED0323FDD3DEC8542584FA0E6CB990FAD5878DB34E9BEDDC95B65D22790C108E218407ED7F7D686657BAB5A28EF81C2E24985B56E37D9934E195A38A835CC02CEE8EBA2F56C87663E332976F5A3720DACA120BCD3DF0AEF6FD78582EBFCEE6D05E06172A871EAB0E8F5FC22DDB600F541B87CF8E147358374406"
+const sigHex = "99394F1D15924C0374CF80C7274CD9F232903A6423D9327156F69743EAEF03E1EFEDFDA8474C97F6570D9EF53C6CE2AE2BA68D01FFF9AA82068214BCD775B95CC297DDC38A63741AB3166B58275E0FB728D26DB18A2C3F14B621CF3863F8648B3149FE896348BE73D37E2F06E6E26C84C044984C09C658300B58EC2383E3B0A1F1390D62B772A69F37B5"
+
+var mISO = []byte("abcdbcdecdefdefgefghfghighijhijkijkljklmklmnlmnomnopnopqo")
+var sigISO []byte
+var qISO []byte
+var idISO []byte
+var svISO signerVerifier
+
+// Test our signer using the values specified in ISO/IEC 14888-2:2008
+func TestSignerISO(t *testing.T) {
+	initializeISOValues(t)
+
+	// The test vector specifies that h(W||M) use SHA-1
+	h := hash
+	hash = sha1Hash
+
+	// Instead of a true random value for r, use the value from the test vector
+	rn := randomNumbers
+	randomNumbers = hardcodedRandomISO
+
+	// restore default function definitions after the test
+	defer func() {
+		hash = h
+		randomNumbers = rn
+	}()
+
+	encodedSig, err := svISO.Sign(qISO, mISO)
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	sig, err := util.Base64DecodeForJWT(encodedSig)
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	if !bytes.Equal(sig, sigISO) {
+		t.Fatal("Signature does not match expected value")
+	}
+}
+
+// Test our verifier using the values specified in ISO/IEC 14888-2:2008
+func TestVerifierISO(t *testing.T) {
+	initializeISOValues(t)
+
+	// The test vector specifies that h(W||M) use SHA-1
+	h := hash
+	hash = sha1Hash
+
+	// The test vector formats Id with PSS. Instead of implementing this
+	// encoding ourselves, we hardcode the value for G given in the standard
+	ep := encodePKCS1v15
+	encodePKCS1v15 = pssEncodedId
+
+	// restore default function definitions after the test
+	defer func() {
+		hash = h
+		encodePKCS1v15 = ep
+	}()
+
+	encodedSigISO := util.Base64EncodeForJWT(sigISO)
+	ok := svISO.Verify(encodedSigISO, idISO, mISO)
+
+	if !ok {
+		t.Fatal("signature verification failed")
+	}
+}
+
+func initializeISOValues(t *testing.T) {
```

Comment:
This is minor and non-blocking, but I think it might be nice to change this to an init function:

```suggestion
func init() {
```

That way it will be called automatically once rather than us having to do it manually in each test. The disadvantage is that it won't have access to the `testing.T` so we'll have to panic on any errors. WDYT?

Reply:
You can use `TestMain` so as not to pollute the `init` space, https://pkg.go.dev/testing#hdr-Main

Reply:
I was on the fence about this and you've convinced me that `init()` is better -- it's definitely more idiomatic and `t.Fatal` isn't really any better than `panic` so I'm good with it!